### ISSUE TITLE
Perform Gaussian fit if more than 3 wavelengths

### ISSUE THREAD

**Issue by mgalloy** [NCAR/ucomp-pipeline#259]
Instead of always doing an analytical Gaussian "fit", the pipeline should do an actual Gaussian fit if there are more than 3 wavelengths in the file.

The code to perform the fit, per pixel, is:

``` IDL
fit = gaussfit(wavelengths, intensities, coefficients, nterms=4)
```

#### Questions

- [ ] Should I save coefficients from the fit somewhere? That would require another extension in the file because it would be 4 coefficients per pixel of the image.
- [ ] What is the convergence criteria? chi-squared less than some threshold?

**Comment by bberkeyU** [comment]
How did we handle fitting in comp? Was it always the analytical gauss?

**Comment by bberkeyU** [comment]
At least for the initial testing I think having the fits and chisq saved for each pixel would be helpful. 

My naive starting point for the threshold would be picking a chisq (or other metic) value that passes a similar set of pixels as the pipeline currently passes with analytical gauss metic

**Comment by detoma** [comment]
I had already opened a ticket for this #253. A better way to handle the extra points is not to always make a least square fit if there are more the 3 wavelengths, but have an option if there are more than 3 wavelengths. There may be cases when we may want to revert to the analytic fit. 

The new wavelength positions, if there are more then 3, may not be equally spaced, so this is something the pipeline needs to handle.

**Comment by mgalloy** [comment]
I am not sure what you mean about the pipeline handling the wavelengths not being equally spaced. The wavelengths will be input as into `gaussfit` — they could be anything as long as they are given to `gaussfit`.

**Comment by mgalloy** [comment]
I have a run in `process.latest` for 20240330. There are new extensions in the level 2 files that give parameters of the fit, by pixel:

```
...
 13  Fit coefficients           science  -----  -----  -----
 14  Fit chi-squared            science  -----  -----  --—-
```

They are arrays giving the coefficients and chi-squared arrays, by pixel:

``` IDL
IDL> help, coeffs, chisq
COEFFS          FLOAT     = Array[1280, 1024, 4]
CHISQ           FLOAT     = Array[1280, 1024]
```

**Comment by detoma** [comment]
I like to have an option in the pipeline to do the analytical gaussian fit or the least square fit for the cases with 5 or more wavelengths.  I think the analytical gaussian fit assume the wavelengths are equally spaced but the 5 or more wavelength positions may not be equally spaces. We need to put some logic on how to select the proper 3 wavelengths for the analytical gaussian fit in case we have more than 3 points. I think the way to do it is to find the two closest wavelengths to the blue and red used for the waves. This are not necessarily the three center wavelengths.

In other words, we cannot simply choose the center 3 wavelengths for the analytical gaussian fit if we have more than three points.

**Comment by mgalloy** [comment]
OK, so the wavelengths only matter if we are using the analytic solution with more than 3 wavelengths? Are the following statements all true?

- If there are only 3 wavelengths, we obviously do an analytic solution with them.
- If there are more than 3 wavelengths and we have the fit option set, we use all the wavelengths in an actual Gaussian fit.
- If there are more than 3 wavelengths and we have the fit option not set, we have some scheme to pick which 3 wavelengths to use in the analytic solution.

**Comment by bberkeyU** [comment]
The LOS velocity, Line width, and Noise mask are 0 (or NaNs) across the whole array in the L2 files in current process.latest run.  Was this expected?

**Comment by mgalloy** [comment]
Yes, I haven't figured out how to actually calculate anything yet (see questions at the top). But there is chi-squared array for these fits that is interesting to look at in that run.

**Comment by bberkeyU** [comment]
@mgalloy I agree chi-squared looks weird.  I assumed it would better match the corona.   

![image](https://github.com/NCAR/ucomp-pipeline/assets/22605316/d621cc0f-2788-4d70-bd5d-f6e9160ed248)
![image](https://github.com/NCAR/ucomp-pipeline/assets/22605316/a9a4198e-df37-4023-876e-c5b4399741b2)

**Comment by bberkeyU** [comment]
@mgalloy I agree, but to first order it should show were the corona is.

**Comment by bberkeyU** [comment]
For the fit masking, I think we can look at the first term of the fit.

If fit[0] > 0, we get a mask that matches reasonably well with the locations with good signals in the peak intensity plots.

**Comment by bberkeyU** [comment]
Can we pull the LOS and line widths directly from the Gaussian fit results fit[1] and fit[2], with appropriate corrections to convert from nm to km/s?

For 1074.7 LOS, I think this would look like 
LOS = (fit[1]-1074.7)*299792.458/1074.7

We need to perform the conversion for the line width, the idl gauss routine reports in stdev, not fwhm, and so on. 
line_width = fit[2]-1074.7)*299792.458/1074.7 * 2.355

**Comment by detoma** [comment]
Ben, please let us work on this. I just sent an email. Yes, the conversion will be done.

**Comment by detoma** [comment]
There are several issues that need to be taken care of to allow the least-square fit to properly work in the automatic pipeline. The strategy Mike and I have agreed on is the following:

- The analytical gaussian fit will be performed for all lines. For the lines that have only three points, this will be the end of the processing and nothing will change compared to before. For the lines that have more than three points, we will identify the wavelength closer to the "nominal" ones used in the three points observations and then perform the analytic gaussian fit on those three points. This is because using the center three points does not always work well, especially if we oversample the line and the three center points are very close together.

- The analytic gaussian fit will be used as an initial guess for the least square gaussian fit with 4 terms. This is because the least square fit does not always give good fits, especially in pixels with low signal. We found unrealistic gaussian profiles see examples below). To further avoid failing of the least square fitting procedure, we need to consider how to select the points for the fit, e.g. eliminating points below a certain threshold or negative. Also, the chi square is not normalized and results in very low values in the outer for where the intensity is low, even if the fit is poor. Mike will normalize the chi square so we can use it as an indicator of how good the fit really is

- There will be a second "custom" option for lines with more than three points, where we specify the wavelengths to use in the fit. This is because there may be observing programs where we want to esclude some wavelength positions that are not reliable.

- Finally, since the analytical gaussian fit equation is different from the one used in IDL gaussfit, we need to make sure that all quantities are consistent and we initialize the fit with the proper coefficients. The final products will be the same, regardless of the fitting technique: peak intensity, FWHM, and doppler velocity.

After all this is implemented, we need to verify if the peak intensity, velocity and FWHM look good or if the noise in the outer fov is too large. I worry about noise in the outer fov especially for the weaker lines. We may choose the 3-term fit, if more stable. It will be interesting to compare the peak intensity, velocity,  and FWHM  for the 3-term and 4-term fit.

Example of fits:

<img width="698" alt="Screenshot 2024-04-16 at 6 31 08 PM" src="https://github.com/NCAR/ucomp-pipeline/assets/12946149/5b42d0b9-7915-4d3e-8dc2-97738e74a64d">

The white is the analytical fit, the blu the least square with 3 terms, and the red the least square with 4 terms. The 4 terms fit should provide a more realistic peak intensity and line width and a better fit in the wings but fails badly in some pixel, as shown above.

<img width="699" alt="Screenshot 2024-04-16 at 6 50 50 PM" src="https://github.com/NCAR/ucomp-pipeline/assets/12946149/b0b9f6b5-7e45-47ad-94eb-21bcce23add1">

Another failed fit.

<img width="698" alt="Screenshot 2024-04-16 at 6 40 16 PM" src="https://github.com/NCAR/ucomp-pipeline/assets/12946149/9eba0465-5013-4467-bd12-51e09c081e68">
<img width="693" alt="Screenshot 2024-04-16 at 6 51 49 PM" src="https://github.com/NCAR/ucomp-pipeline/assets/12946149/8d33e1cd-6083-45fc-a4e6-2d48c6962a4c">

Two examples where the least square fit worked well.

**Comment by mgalloy** [comment]
Doing the analytical "fit" takes about 2-5 seconds for a level 2 file. Doing the actual Gaussian fit takes 7-15 minutes per level 2 file.

**Comment by detoma** [comment]
Would rewrite this in C speed it up?

The 7-15 minutes are for a fitting all the pixels or only the pixels that pass the velocity criteria?

Criteria need to be revisited for 5-7 points data. I do not think we want to use negative values. The current criteria only apply to the three center points.

**Comment by mgalloy** [comment]
There are several things that we can do to speed it up:

1. mask pixels so we don't have to fit all of them like it is doing now
2. using the `ESTIMATES` keyword, which we planned to do anyway to get a better fit, would allow the fit to converge faster
3. fitting individual pixels is an "embarrassingly parallel" operation, so we could run on multiple cores

Remember that the waves program is a lot of the files for a day and would not use the fit.

**Comment by detoma** [comment]
Below is an example when the least square gaussian fit uses the far red and blue points that are too low and find an unrealistic gaussian fits. The blue curve is a fit with 3 terms and the red curve a fit with 4 terms. They are both bad fits to the data. We need to exclude data points with intensity below a threshold (0.1 ppm??). 

<img width="695" alt="Screenshot 2024-04-18 at 1 38 45 PM" src="https://github.com/NCAR/ucomp-pipeline/assets/12946149/34635f25-da16-463e-be44-78040c6aebb5">


Below is the same plot excluding the first and last point which are below 0.1ppm. The 3 term fit is now good, the 4 terms fit is bad.

<img width="698" alt="Screenshot 2024-04-18 at 1 48 29 PM" src="https://github.com/NCAR/ucomp-pipeline/assets/12946149/c78e0dc7-3701-4c68-ba40-56b203b41d69">


Below is the same plot excluding the first and last point and using the estimate from the analytical gaussian fit to help converging. The 3 term fit is good, the 4 terms is still bad.

<img width="699" alt="Screenshot 2024-04-18 at 1 55 05 PM" src="https://github.com/NCAR/ucomp-pipeline/assets/12946149/0f34f7ec-fd12-4612-854e-fec596e647b5">

Note: In the analytical fit I used the wavelengths closets to the nominal 3-points position and not the 3 center wavelengths.

Conclusions:
- We need to exclude points with intensity below 0.1ppm. This means not all pixels will use the same number of wavelengths to compute the gaussian, which is not ideal.
- Adding the keyword "estimates" with values from the analytic fit works well and speed up the fit.
- We may not be able to use the 4-term fit that includes a background term.

**Comment by bberkeyU** [comment]
How do we calculate lambda_zero is this 1074.7 for our favorite wavelength region?

**Comment by detoma** [comment]
We do not calculate lambda_zero. Lambda_zero is the nominal center wavelength and is our reference. The analytical gaussian fit returns doppler_shift which is the display in nm of the peak of the fitted gaussian from the nominal center wavelength position.

Because we are not always well centered on the line, the velocities are later renormalized by subtracting the "rest wavelength" computed by averaging the median velocity of the east and west coronal pixels above a certain threshold.

**Comment by mgalloy** [comment]
I think I have reasonable runs for the analytic Gaussian "fit", the 3-term fit, and the 4-term fit in:

- `process.analytic-gaussian`
- `process.3-term`
- `process.4-term`

**Comment by mgalloy** [comment]
I reprocessed the `analytic-gaussian` run to have the same distortion and be comparable to the `3-term` and `4-term` runs.

**Comment by mgalloy** [comment]
I processed 20220115 and 20220116 with both analytic, 3-term, and 4-term fits. Results in:

- `process.analytic-gaussian`
- `process.3-term`
- `process.4-term`

**Comment by bberkeyU** [comment]
Looking at 20220115.191933.ucomp.1074.l2.fts

I find the 3 and 4 term fits have identical LOS pixels values. But both have a ~ 30 km/s offset between the analytical and 3/4 point fits.    Is this what we expected?

**Comment by bberkeyU** [comment]
The noise masking appears to fail for  20220115.191933 fits file.  In the analytic-gaussian the noise mask values range from 0-1, instead of a binary good which I expected from the code.      And the line width appears null in the 3/4 term fits so the mask is also null for these data.

![image](https://github.com/NCAR/ucomp-pipeline/assets/22605316/12efb3cf-d10a-4a7f-91d7-00cbebf39b85)

**Comment by detoma** [comment]
I looked at a few random 1074 images for the 3 days and compared peak intensity, velocity and line width. The images in the directories are identical (min and max of the difference image is zero). I do not see a 30km/s offset. 

I noticed the peak intensity has noise in the outer fov. I am not sure what is causing it.

<img width="1279" alt="Screenshot 2024-06-05 at 2 40 20 PM" src="https://github.com/NCAR/ucomp-pipeline/assets/12946149/a0f8619d-3fd9-4e23-8847-427f9086510b">

**Comment by mgalloy** [comment]
Questions about the current runs:

- [ ] Confirm settings, especially for the 2022 runs (are the fits really being done, or is it falling back to analytic Gaussian "fit").
- [ ] How is `RSTWVL` computed? why 0.0 so much? or -33?
- [ ] Noise mask for `analytic-gaussian` had values between 0 and 1.
  - for example, `20220115.191933`
  - I don't see this, not sure of what the discrepancy is.
- [ ] Are 3- and 4-term fits identical except mask? why?

**Comment by bberkeyU** [comment]
This doesn't add any new information from yesterday's meeting.  Three versions of the fits seem identical.

I put together some Python code to look at the LOS value for all the pixels 2380 pixels in a ring at height =300 (pixels)  for every 1074 5 point data reprocessed in the 'process.analytic-gaussian', 'process.4-term' processing directories.

After correcting the pixel values by adding the rest wavelength reported in the header.  The 3 and 4-term fits are identical.   The differences between the analytic and fit are very small, less than .0005 km/s.    Most of this difference seems to be from a DC offset (maybe noise in the rest wavelength calculation) that changes per fits file.  Within a single file, the differences between the fits are closer to 1e-5 km/s.

**Comment by bberkeyU** [comment]
@detoma @jburkepile This is just a reminder that on 2022-02-02, we moved the field lens to its correct location (closer to the lyot stop). This should have changed the background (I forget if it did). In today's meeting, we talked about adding some February dates to the Gaussian data checking; it would be interesting to compare the before-and-after field lens moves.

**Comment by jburkepile** [comment]
Thanks for checking the change to the field lens location Ben and adding it to the ticket. Looks like you have data before and after the field lens change on Feb 2. 
It looks like the stray light begins to diminish on Feb 8 and Feb 9 and by Feb 10, 2022 the data look good. It would be good to have Mike run Gaussian fits (analytic, 3- and 4-parameter) for a sample of days. How about reprocessing:  Feb 1, Feb 2, Feb 7, Feb 10, Feb 14, 2022
 Are 5 days sufficient?

**Comment by mgalloy** [comment]
So not Jan 15, 16, and 19? or 20240409?

**Comment by detoma** [comment]
We have paused this issue for a while. The analysis of the widths in 1074 and 1079 shows systematic differences. This indicates we need to improve our fit.

Unfortunately we stopped taking 5-points profiles in Feb 2022 and the data taken in 2021 and Jan 2022 were still very noisy. When we reopen, we will go back to take 5-point synoptic data to have enough points across the line that allow to use a gaussian 4-term fit.

We identified an issue with the 4-term gaussfit in IDL which often does not converge even with a good initial guess. This is likely a problem with how the least-square gaussian fit is implemented. To use a 4-term fit, it will require Mike to write a more robust fit instead of using the standard gaussfit. 

Examples of fits are given below.

**Comment by detoma** [comment]
I worked at improving the initial guess: I fixed a bug in the width initial guess and now use the 3-term gaussfit coefficients as the initial guess for the 4-term gaussfit. This seems to have fixed the issue with the anomalous fits shown above. I will push the changes in the pipeline. As a reminder, I introduced the initial guess in gaussfit because without an initial guess gaussfit failed to converge often. An initial guess is necessary.

Remaining issues with the 3-terms 4-term gaussian fits:

- Some fits are bad simply because the input data is very noisy. We should use the chi-square information and set to NaN pixels for which the chi squared is too high. This is a useful additional criterium for our gaussian fit. For example, low intensity profiles could be accepted only if the chi-square is good. 

- At times the chi-square for the 4-term is worse than the chi-square for the 3-terms. In these cases, we have three options:
1)accept the 4-term fit as long as the chi-square is within our limits
2)set that pixel to NaN in the 4-term velocity/width 
3)adopt the 3-term values
I think option 1) is preferable.

- When intensities are very low, we can get a negative background in the 4-terms fit. While this scenario is technically possible if we over-estimate the continuum subtraction, fits that showed a negative background often looked strange. I cannot make conclusions by checking only a few profiles but it seems safer to remove these points by setting the velocity/width to NaN when the background term is negative or by setting  the minimum acceptable negative background to a small negative number, something like -0.1. We should count how many pixels in an image have negative background because this indicates a problem and should check if this correlates to spilage or bad sky. Even better, we should add an extension to the 4-terms fit that record the background level found by the fit. The background should be smooth over the detector. 

- We are not passing an error for the intensities to gaussfit. This could/should be added but require testing and care.

**Comment by detoma** [comment]
Comments from Joan:

<div class="gE iv gt" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none; font-size: 0.875rem; padding: 20px 0px 0px; cursor: auto;">
Joan Burkepile <iguana@ucar.edu> | Joan Burkepile <iguana@ucar.edu> | May 22, 2025, 9:06 AM |   |  
-- | -- | -- | -- | --
Joan Burkepile <iguana@ucar.edu>
to me, Michael | to me, Michael
to me, Michael

</div><div id=":487" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><div class="KKSLrd"></div><div class="qQVYZb"></div><div class="utdU2e"></div><div class="lQs8Hd" jsaction="SN3rtf:rcuQ6b" jscontroller="i3Ohde"></div><div class="wl4W9b" jsaction="LNSvUb:.CLIENT;xSdBYb:.CLIENT;CDWmBe:.CLIENT;EtHLdc:.CLIENT;pQnh7:.CLIENT;pKHw7e:.CLIENT;Z03mxd:.CLIENT;NZLNxf:.CLIENT;bXglpe:.CLIENT;mzh2Bc:.CLIENT" style="display: flex;"></div></div><div class="" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><div class="aHl" style="margin-left: -30px; caret-color: rgb(34, 34, 34); color: rgb(34, 34, 34); font-family: &quot;Google Sans&quot;, Roboto, RobotoDraft, Helvetica, Arial, sans-serif; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"></div><div id=":45z" tabindex="-1" style="caret-color: rgb(34, 34, 34); color: rgb(34, 34, 34); font-family: &quot;Google Sans&quot;, Roboto, RobotoDraft, Helvetica, Arial, sans-serif; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"></div><div id=":484" class="ii gt" jslog="20277; u014N:xr6bB; 1:WyIjdGhyZWFkLWY6MTc5NjQ0NTE3Mjk1MDAyNjY0Mnxtc2ctZjoxODMyODMzNjYzNTczMDY0OTQxIl0.; 4:WyIjbXNnLWY6MTgzMjgzMzY2MzU3MzA2NDk0MSIsbnVsbCxudWxsLG51bGwsMSwwLFsxLDEsMF0sNjQzLDM3NDgsbnVsbCxudWxsLG51bGwsbnVsbCxudWxsLDEsbnVsbCxudWxsLFswXSxudWxsLG51bGwsbnVsbCxudWxsLG51bGwsbnVsbCwwLDBd" style="direction: ltr; margin: 8px 0px 0px; padding: 0px; position: relative; font-size: 0.875rem; overflow-x: hidden; caret-color: rgb(34, 34, 34); color: rgb(34, 34, 34); font-family: &quot;Google Sans&quot;, Roboto, RobotoDraft, Helvetica, Arial, sans-serif; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><div id=":485" class="a3s aiL " style="direction: ltr; font-style: normal; font-variant-caps: normal; font-weight: normal; font-stretch: normal; font-size: small; line-height: 1.5; font-family: Arial, Helvetica, sans-serif; font-size-adjust: none; font-kerning: auto; font-variant-alternates: normal; font-variant-ligatures: normal; font-variant-numeric: normal; font-variant-east-asian: normal; font-variant-position: normal; font-feature-settings: normal; font-optical-sizing: auto; font-variation-settings: normal; overflow: auto hidden; position: relative;"><u></u><div><p>Dear Giuliana:</p><p>This is excellent work! It is obvious you put a tremendous amount of effort and thought into this. Great job! I have summed up your action items in order to assess how much time Mike needs to implement all this. He also has many items on his plate. He and I are meeting at noon and can talk about it.<br></p><p>1) We should use the chi-square information and set to NaN pixels for which the chi squared is too high. This is a useful additional criterion for our<span class="Apple-converted-space"> </span><span class="il">gaussian</span><span class="Apple-converted-space"> </span>fit. For example, low intensity profiles could be accepted only if the chi-square is good.</p><p>2) At times the chi-square for the 4-term is worse than the chi-square for the 3-terms. When this happens accept the 4-term fit as long as the chi-square is within our limits</p><p>3) count how many pixels in an image have negative background because this indicates a problem and should check if this correlates to spillage or bad sky. Even better, we should add an extension to the 4-terms fit that record the background level found by the fit.</p><p>4) Remove negative background points by setting the velocity/width to NaN when the background term is negative or by setting the minimum acceptable negative background to a small negative number, something like -0.1.</p><p>5) We are not passing an error for the intensities to gaussfit. This could/should be added but require testing and care.</p><p>Thanks so much for doing all this!!</p><p>all the best,</p><p>joan<br></p><p></p></div></div></div><br class="Apple-interchange-newline"></div><br class="Apple-interchange-newline">
